src/trajectory.jl

```diff
@@ -718,6 +770,55 @@ function build_tree(
     end
 end
 
+"""
+Iteratively build a tree for a given depth `j`.
+"""
+function build_tree(
+    rng::AbstractRNG,
+    nt::IterativeNUTS{S,C,I,F},
+    h::Hamiltonian,
+    z::PhasePoint,
+    sampler::AbstractTrajectorySampler,
+    v::Int,
+    j::Int,
+    H0::AbstractFloat,
+) where {I<:AbstractIntegrator,F<:AbstractFloat,S<:AbstractTrajectorySampler,C<:AbstractTerminationCriterion}
+    ileaf_max = 2^j
+    state_cache_size = j
+
+    z′, state′ = build_one_leaf_tree(nt, h, z, sampler, v, H0)
+    j == 0 && return state′.tree, state′.sampler, state′.termination
+
+    # TODO: allocate state_cache in `transition` and reuse for all subtrees
+    state_cache = Vector{typeof(state′)}(undef, state_cache_size)
```

Comment:
Just to check my understanding. The reason we have the `-1` in L727 and L815 is because we are using the `state_cache` in a binary format, thus `state_cache[end]` is the least significant digit.

Reply:
The reverse iteration is because we combine a subtree with the most recent subtree of the same size. And that produces a subtree that must be combined with the next most recently merged subtree, which will always be the previous entry in the cache. `subtree_cache_combine_range` isn't very intuitive, but it gives us this range, as checked against a simplified recursion. I'll add a suitable test.